File: src/hedera-integration.ts
// Hedera SDK imports
import {
  Client,
  AccountId,
  PrivateKey,
  TopicId,
  TopicCreateTransaction,
  TopicMessageSubmitTransaction,
} from "@hashgraph/sdk";

export interface AuditEntry {
  id: string;
  timestamp: string;
  agentId: string;
  action: "SWARM_START" | "SWARM_COMPLETE" | "AGENT_BIRTH" | "AGENT_DEATH" | "AGENT_COMPLETE";
  payload: string;
  signature?: string;
  topicId?: string;
  sequenceNumber?: number;
}

export interface PQCSigningResult {
  signature: string;
  publicKey: string;
  timestamp: string;
}

export class HederaIntegration {
  private client: Client;
  private topicId?: TopicId;
  private operatorId?: AccountId;
  private operatorKey?: PrivateKey;

  constructor(network: "testnet" | "mainnet" = "testnet") {
    const envVar =
      network === "testnet"
        ? "HEDERA_OPERATOR_ID"
        : "HEDERA_MAINNET_OPERATOR_ID";
    const keyVar =
      network === "testnet"
        ? "HEDERA_OPERATOR_KEY"
        : "HEDERA_MAINNET_OPERATOR_KEY";

    const operatorId = process.env[envVar];
    const operatorKey = process.env[keyVar];

    if (operatorId && operatorKey) {
      this.operatorId = AccountId.fromString(operatorId);
      this.operatorKey = PrivateKey.fromString(operatorKey);
      this.client =
        network === "mainnet"
          ? Client.forMainnet().setOperator(this.operatorId, this.operatorKey)
          : Client.forTestnet().setOperator(this.operatorId, this.operatorKey);
    } else {
      this.client =
        network === "mainnet" ? Client.forMainnet() : Client.forTestnet();
    }
  }

  async createAuditTopic(description: string): Promise<string> {
    if (!this.operatorId || !this.operatorKey) {
      throw new Error(
        "Operator credentials required. Set HEDERA_OPERATOR_ID and HEDERA_OPERATOR_KEY",
      );
    }

    const transaction = new TopicCreateTransaction().setTopicMemo(
      `Swarm Spawner Audit: ${description}`,
    );

    const response = await transaction.execute(this.client);
    const receipt = await response.getReceipt(this.client);

    if (receipt.topicId) {
      this.topicId = receipt.topicId;
      return this.topicId.toString();
    }
    throw new Error("Failed to create topic");
  }

  async logSwarmStart(data: {
    taskCount: number;
    strategy: string;
    timestamp: string;
  }): Promise<AuditEntry> {
    const entry: AuditEntry = {
      id: `audit-${Date.now()}`,
      timestamp: data.timestamp,
      agentId: "swarm-coordinator",
      action: "SWARM_START",
      payload: JSON.stringify(data),
    };

    return this.submitAuditEntry(entry);
  }

  async logSwarmComplete(data: {
    successCount: number;
    failureCount: number;
    duration: number;
    auditEntries: AuditEntry[];
  }): Promise<AuditEntry> {
    const entry: AuditEntry = {
      id: `audit-${Date.now()}`,
      timestamp: new Date().toISOString(),
      agentId: "swarm-coordinator",
      action: "SWARM_COMPLETE",
      payload: JSON.stringify(data),
    };

    return this.submitAuditEntry(entry);
  }

  async signAndLog(
    agent: { id: string; task: string; status: string },
    keyPair: { publicKey: string; privateKey: string },
  ): Promise<AuditEntry> {
    const payload = JSON.stringify({
      agentId: agent.id,
      task: agent.task,
      status: agent.status,
    });
    const pqcResult = await this.signWithPQC(payload, keyPair);

    const entry: AuditEntry = {
      id: `audit-${Date.now()}-${agent.id}`,
      timestamp: new Date().toISOString(),
      agentId: agent.id,
      action: "AGENT_COMPLETE",
      payload,
      signature: pqcResult.signature,
    };

    return this.submitAuditEntry(entry);
  }

  /**
   * @deprecated Use PQCIdentityManager for real ML-DSA-65 signing.
   * This legacy method is retained only for backward compatibility
   * with v0.1.0 consumers. It provides NO cryptographic security.
   */
  async signWithPQC(
    payload: string,
    _keyPair: { publicKey: string; privateKey: string },
  ): Promise<PQCSigningResult> {
    console.warn(
      "[Hedera] signWithPQC() is deprecated and provides no security. Use PQCIdentityManager instead.",
    );
    return {
      signature: "(deprecated-no-crypto)",
      publicKey: _keyPair.publicKey,
      timestamp: new Date().toISOString(),
    };
  }

  /**
   * @deprecated Use PQCIdentityManager.verify() for real ML-DSA-65 verification.
   * This method always returns false to prevent false-positive trust.
   */
  async verifyPQCSignature(
    _payload: string,
    _signature: string,
    _publicKey: string,
  ): Promise<boolean> {
    console.warn(
      "[Hedera] verifyPQCSignature() is deprecated. Use PQCIdentityManager.verify() instead.",
    );
    return false;
  }

  private async submitAuditEntry(entry: AuditEntry): Promise<AuditEntry> {
    if (!this.topicId || !this.operatorId || !this.operatorKey) {
      console.log("[Hedera] Audit entry (no topic configured):", entry);
      return entry;
    }

    try {
      const transaction = new TopicMessageSubmitTransaction({
        topicId: this.topicId,
        message: JSON.stringify(entry),
      });

      const response = await transaction.execute(this.client);
      const receipt = await response.getReceipt(this.client);

      entry.topicId = this.topicId.toString();
      entry.sequenceNumber = receipt.topicSequenceNumber?.low ?? 0;
    } catch (error) {
      console.error("[Hedera] Failed to submit audit entry:", error);
    }

    return entry;
  }

  async close(): Promise<void> {
    await this.client.close();
  }
}

File: src-rs/src/hedera_integration.rs
use serde::{Deserialize, Serialize};
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuditEntry {
    pub id: String,
    pub timestamp: String,
    pub agent_id: String,
    pub action: String,
    pub payload: String,
    pub signature: Option<String>,
    pub topic_id: Option<String>,
    pub sequence_number: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PQCSigningResult {
    pub signature: String,
    pub public_key: String,
    pub timestamp: String,
}

pub struct HederaIntegration {
    network: String,
    topic_id: Option<String>,
}

impl HederaIntegration {
    pub fn new(network: &str) -> Self {
        Self {
            network: network.to_string(),
            topic_id: None,
        }
    }

    pub async fn log_swarm_start(&self, data: serde_json::Value) -> AuditEntry {
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis();

        AuditEntry {
            id: format!("audit-{}", timestamp),
            timestamp: chrono_lite_timestamp(),
            agent_id: "swarm-coordinator".to_string(),
            action: "SWARM_START".to_string(),
            payload: serde_json::to_string(&data).unwrap_or_default(),
            signature: None,
            topic_id: None,
            sequence_number: None,
        }
    }

    pub async fn log_swarm_complete(&self, data: serde_json::Value) -> AuditEntry {
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis();

        AuditEntry {
            id: format!("audit-{}", timestamp),
            timestamp: chrono_lite_timestamp(),
            agent_id: "swarm-coordinator".to_string(),
            action: "SWARM_COMPLETE".to_string(),
            payload: serde_json::to_string(&data).unwrap_or_default(),
            signature: None,
            topic_id: None,
            sequence_number: None,
        }
    }

    pub async fn sign_and_log(
        &self,
        agent_id: &str,
        task: &str,
        status: &str,
        key_pair: &super::PQCKeyPair,
    ) -> AuditEntry {
        let payload = serde_json::json!({
            "agentId": agent_id,
            "task": task,
            "status": status,
        });

        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis();

        AuditEntry {
            id: format!("audit-{}-{}", timestamp, agent_id),
            timestamp: chrono_lite_timestamp(),
            agent_id: agent_id.to_string(),
            action: "AGENT_COMPLETE".to_string(),
            payload: payload.to_string(),
            signature: Some("pqc-signature-placeholder".to_string()),
            topic_id: None,
            sequence_number: None,
        }
    }

    pub async fn mint_credential(
        &self,
        agent_id: &str,
        metadata: serde_json::Value,
    ) -> String {
        println!("[Hedera] Minting credential for agent {}", agent_id);
        format!("credential-{}-{}", agent_id, SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis())
    }
}

fn chrono_lite_timestamp() -> String {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis();
    format!("{}", now)
}

pub mod mock {
    pub fn verify_pqc_signature(_payload: &str, _signature: &str, _public_key: &str) -> bool {
        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_audit_entry_creation() {
        let hedera = HederaIntegration::new("testnet");
        let data = serde_json::json!({"taskCount": 5, "strategy": "parallel"});
        let entry = hedera.log_swarm_start(data).await;
        
        assert_eq!(entry.action, "SWARM_START");
        assert_eq!(entry.agent_id, "swarm-coordinator");
    }
}